src/Features/CSharp/Portable/Copilot/CSharpCopilotNotImplementedMethodFixProvider.Analyzer.cs

```diff
@@ -0,0 +1,157 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.MethodImplementation;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Copilot;
+
+internal sealed partial class CSharpCopilotNotImplementedMethodFixProvider
+{
+    private static class DocumentAnalyzer
+    {
+        private const int MaxMethodLength = 1024;
+        private const int ContextLineCount = 10;
+
+        public static async Task<MethodImplementationProposal?> AnalyzeDocumentAsync(Document document, SyntaxNode throwNode, CancellationToken cancellationToken)
+        {
+            // Get the semantic model and syntax root
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // Find the containing method declaration
+            var methodDeclaration = throwNode.Parent.FirstAncestorOrSelf<MethodDeclarationSyntax>();
+
+            // Get symbol information
+            var methodSymbol = semanticModel.GetRequiredDeclaredSymbol(methodDeclaration, cancellationToken);
+
+            // Find references
+            var references = await FindReferencesAsync(methodSymbol, document, cancellationToken).ConfigureAwait(false);
+            var referenceCount = references.Sum(r => r.Locations.Count());
+
+            // Get top 2 surrounding code snippets
+            var topReferences = references
+                .SelectMany(r => r.Locations)
+                .OrderBy(l => l.Location.SourceSpan.Length)
+                .Take(2)
+                .Select(async l =>
+                {
+                    var refDocument = document.Project.Solution.GetDocument(l.Document.Id);
+                    if (refDocument == null)
+                        return null;
+                    var refRoot = await refDocument.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
this will get (and possibly drop) the root for all these documents multiple times (even if finding the same ref in multiple docs).

What you want to do is a GroupBy on the result to group by document.  Then lookup all the spans of interest, then find the containing BaseProperty or BaseMethod (and do a Distinct to ensure fhtat if multiple refs are int he same method you only process them once).